---
title: "miRDriver"
author: "Banabithi Bose, Matt Moravec,  Serdar Bozdag"
date: "9/9/2020"
output:      
    BiocStyle::html_document:
        toc: true
        toc_depth: 3  
        number_sections: true
        theme: united 
r_packages: rmarkdown
vignette: > 
    %\VignetteIndexEntry{miRDriver_new} 
    %\VignetteEngine{knitr::rmarkdown} 
    %\VignetteEncoding[UTF-8]{inputenc}
---


# <a id = "Introduction"></a> Introduction
miRDriver is a computational tool that infers copy number aberation-bsed miRNA-gene interactions in cancer. miRDriver has four computational steps. In the first step we find frequently aberrated regions that have been retrieved from the GISTIC tool (https://cloud.genepattern.org/gp/pages/index.jsf?lsid=urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00125:6.15.28). 
In the second step, for each GISTIC region we computed differentially expressed (DE) genes between frequently aberrated and non-aberrated patient groups. In the third step, we retrieved DE genes and miRNAs that reside in aberrated regions (i.e. cis genes and cis miRNAs) and retrieved DE genes that are outside of aberrated regions (i.e. trans genes). In the last step, we employed a LASSO-based regression model to compute miRNA regulators of the trans genes.


```{r}
library("miRDriver");
```

# <a id = "Example-Data-Sets"></a> Example Data Sets
The datasets included are included to give the user an idea of the type of data that is required to run the pipeline. This includes genomic data (gene expression, copy number variation, transcription factor) and epigenomic data (methylation). The gene expression data can be RNASeq or Microarray data with proper  normalization, althought the code has not been tested on Microarray data. The first phase of miRDriver requires copy number data files that are segmented. The Lasso phase of miRDriver requires gene centric copy number data. The methylation data needs to be gene centric beta values. For further details on how to preprocess the datasets, one should reference the method paper "miRDriver: A tool to infer copy number derived miRNA-Gene networks in cancer". The data sets included in this package are dummy datasets that are used to run the code in the vignette, but are not large enough to produce meaningful results. The original data that the dummy datasets are derivedfrom are included on our github page. Our original data sets are from TCGA (The Cancer Genome Atlas) database. For actual datasets one can use the TCGA (The Cancer Genome Atlas) database or any other resource. When preparing the dummy data sets we chose 10 patients, and cropped each dataset to only include the 10 chosen patients. We also reduced the number of genes and only included the genes that were a part of the chosen group.

## <a id = "mirna-bedfile"></a> mirna bedfile
This is an R dataframe object where the data represents the genomic position of the miRNA's. The rows in this dataset are the chromosomal regions that start with "chr1", the same notation carries forward to all chromosomes. The columns will be “chromosome number”, “start position”, “end position”, and "mirna". The chromosome number will follow the notion of "chr1", the start position and end positions will be integers and the mirna will be a string with the mirna ID.


```{r}
load(system.file("extdata", "mirna_bedfile.Rda", package="miRDriver"))
mirna_bedfile[1:5,]
```

## <a id = "RNA-seq-count-data"></a> RNA seq count data
This is an R matrix object where the data is the RNA Seq count data for each patient and each gene in the study. The count values are normalized using FPKM. The rows of this data set are ENSEMBLE ID's for each gene. The columns are sample id's for the patients in the study.

```{r}
load(system.file("extdata", "Dummy_RnaSeq_Count_Data.Rda", package="miRDriver"))
Dummy_RnaSeq_Count_Data[1:5,1:3]
```

## <a id = "lesions-file"></a> lesions file
This text file contains information about possible aberated genetic segments, "peaks", and the aberation status for each sample. This files is made by the GISTIC tool. The first 9 columns of the GISTIC output the columns will be “Unique.Name, Descriptor, Wide.Peak.Limits, Peak.Limits, Region.Limits, q.values". After the first 9 columns there are sample id's with their aberation status "0", "1", "2". The status level 0 represents there not being a copy number  aberation, status level "1" represents a low level copy number aberation, and "2" represents a high level copy number aberation.

```{r}
x<-read.table(system.file("extdata", "Dummy.gistic.lesions.txt", package="miRDriver"), sep = "\t", header = TRUE)
x[1:5,c(1,10:13)]
```

## <a id = "hg38-genes-coordinate-bedfile"></a> hg38 genes coordinate bedfile
This file is an R dataframe object. It has the coordinates for all the genes in the human genome. There are four columns. They are the chromosome, the start position, the end position and the ENSEMBLE ID. The rows are labeled with the gene id.

```{r}
load(system.file("extdata", "hg38_genes_coord_bedfile.rda", package="miRDriver"))
Genes_coord[1:5,]
```
## <a id = "CNVData"></a> CNVData
This file is a R dataframe object. It containes the copy number variation data. The first column is labeled "sample", and contains the sample id's. The rest of the columns are the the values for the copy number variation for the gene represented with ENSEMBLE ID.

```{r}
load(system.file("extdata", "dummy_CNVData10.Rda", package="miRDriver"))
dummy_CNVData10[1:5,1:3]
```

## <a id = "mirnaData"></a> mirnaData
This file is a R dataframe object. It contains the expression data for the miRNA included in the study. The first column is labeled "sample" and contains the sample id's. The rest of the columns are the data for the miRNA in column title, represented in hsa format.

```{r}
load(system.file("extdata", "dummy_mirnaData10.Rda", package="miRDriver"))
dummy_mirnaData10[1:5,1:3]
```
## <a id = "RSeqData"></a> RSeqData
This file is a R dataframe object. It is the RNA Seq data for the genes and patients included in the study. The first column, labeled sample, has sample id's and the rest of the columns are labeled with the ENSEMBLE ID and have the RSeq values.

```{r}
load(system.file("extdata", "dummy_RSeqData10_200.Rda", package="miRDriver"))
dummy_RSeqData10_200[1:5,1:3]
```
## <a id = "TFData"></a> TFData
This file is a R dataframe object. It has the transription factor (TF) genomic data for the genes in the study. The first column, labeled V1, has the genes in ENSEMBLE ID format and the second column, labeled TF, has the transcription factors in ENSEMBLE ID format.

```{r}
load(system.file("extdata", "dummy_TFData.Rda", package="miRDriver"))
dummy_TFData[1:5,]
```
## <a id = "methylationData"></a> methylationData
This file is a R dataframe object. It contains the DNA methylation data of each gene for each patient. The first column, labeled sample, has the sample id's and the rest of the columns have the methylation values for the gene represented by the ENSEMBLE ID in the column label.

```{r}
load(system.file("extdata", "dummy_methylationData10.Rda", package="miRDriver"))
dummy_methylationData10[1:5,1:3]
```

# <a id = "Using-the-Functions"></a> Using the Functions
The functions are dependent on each other in the sense that user needs to finish the first step of miRDriver in order to start the second step and so on. The following sections go over the how to use each function individually. They are in the order the functions should be run in to complete the pipeline, showing how they are dependent on each other.

## <a id = "Step-1"></a> Step 1
Step one of the miRDriver pipeline finds frequently aberated copy number regions among cancer patients. This step consists of one function, *getRegionWiseGistic()*.

### <a id = "getRegionWiseGistic"></a> getRegionWiseGistic
This function uses the lesions file to find the aberrations for each peak. The lesions file, as stated in the example data sets section, can be created with the GISTIC2.0 module on genepattern.com (https://cloud.genepattern.org/gp/pages/index.jsf?lsid=urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00125:6.15.28).If one uses GISTIC2.0 to make the lesions file, then the lesions file will be labeled all_lesions.conf_90.txt. As one can see in the example code, this file is used in the function as the argument *gisticfile*. The *gistic_bedfile* argument should be either TRUE or FALSE, representing if the user would like the function to make a gistic_bedfile. The *mirdirectory* argument needs to be the file path that the user wants the results to be in. The mirdirectory needs to be the same for all functions in the same analysis.

Arguments:

gisticfile - a character string for the file path of the GISTIC output file

gistic_bedfile - a logical value for whether or not one wants to create a gistic_bedfile (TRUE/FALSE)

mirdirectory - a character string for the file path of the directory one wants the results in

```{r}
getRegionWiseGistic(gisticfile=system.file("extdata", "Dummy.gistic.lesions.txt", package="miRDriver"),gistic_bedfile = TRUE,mirdirectory="~/MIR123")
```

Output: The output from this function can be found in the mirdirectory in mirDriverFold/miRDriver_Step1. This produces the GISTIC bedfile in the GisticResults folder and a file for each peak in the Regionwise_Gistic_Files folder.

```{r}
load("~/MIR123/mirDriverFold/miRDriver_Step1/GisticResults/gistic_bedfile.rda")
region[1:5,]

```

```{r}
x<-data.frame(read.table("~/MIR123/mirDriverFold/miRDriver_Step1/Regionwise_Gistic_Files/Amplification Peak  7.txt", sep = "\t", header = TRUE))
x[1:5,]
```

##<a id = "Step-2"></a> Step 2
Step two of the miRDriver pipeline is to find the differentially expressed (DE) genes. This step uses two functions, *getDifferentiallyExpressedGenes()* and *makingTransAndCisGenes()*.

### <a id = "getDifferentiallyExpressedGenes"></a> getDifferentiallyExpressedGenes
This function finds the differentially expressed genes based off of peak wise files generated in the getRegionWiseGistic function and RNASeq count data. This function can be run in parallel or on one core. The user sets the number of cores in the *ncore* argument or uses the default of two. To run in parallel the argument *parallel* is set to TRUE, to run linearly the argument is set to FALSE. The argumen *RNAcount* needs to be set to the file path for the RNASeq count data. The peakwise files should be in the correct folder to be used with the code,  "~/mirdirectory/mirDriver/miRDriver_Step1/Regionwise_Gistic_Files". In the example mirdirectory represents the *mirdirectory* argument. If the user has run the getRegionWiseGistic function and set the mirdirectory to the same thing in both functions then the peakwise files are in the correct place.

Arguments:

ncore - assuming parallel core use, an integer value for the number of cores user wants to use (default = 2), this argument is ignored if parallel is set to FALSE

RNAcount - a character string for the file path to a R dataframe of RNA seq count

parallel - a logic value (TRUE/FALSE) for whether or not one wants to run the code in parallelE

mirdirectory - a character string for the file path of the directory one wants the results in, must be the same as in all previous functions.


```{r warning=FALSE}
load(system.file("extdata", "Dummy_RnaSeq_Count_Data.Rda", package="miRDriver"))
getDifferentiallyExpressedGenes(RNACount = Dummy_RnaSeq_Count_Data, parallel = FALSE, mirdirectory = "~/MIR123")
```

Output: The output for this function can be found in the mirdirectory in mirDriverFold/miRDriver_Step2/UpDownGeneLibrary. These files are the intermediary files for step two.

```{r}
x<-read.table("~/MIR123/mirDriverFold/miRDriver_Step2/UpDownGeneLibrary/UPgene_Amplification Peak  9.txt", sep = "\t", header = TRUE)
x[1:3,]
```


### <a id = "makingTransAndCisGenes"></a> makingTransAndCisGenes
This function finds the trans and cis genes for each peak. The trans and cis genes will be stored in R dataframe objects located in the file path ~/mirdirectory/miRDriver_Step2/Trans_Cis_Files/peak/. In the file path the mirdirectory is the mirdirectory specified in the *mirdirectory* argument and should be the same as the mirdirectory used in the previous functions. The peak in the file path represents the subdirectories for each peak which will follow the convention "Amplification Peak 1". The *gistic_bedfile* argument needs to be set to a R dataframe object that follows the convention of having four columns, "chromosome", "start position", "end position", and "peak". The chromosome column will have the chromosome number in the form "chr1". The start and end positions will be integers representing the start and end positions on the chromosome. The peak column will have a string representing the peak following the convention of "Amplification Peak 1". The gistic_bedfile can be made in the *getRegionWiseGistic()* function.

Arguments:

Genes_coord_bedfile - A R dataframe object that fits the format of a bedfile for gene coordinates

gistic_bedfile - A R dataframe object that fits the format of a bedfile for the GISTIC results with amplification/deletion peaks with patients

mirdirectory - a character string for the file path of the directory one wants the results in, must be the same as all previous functions

```{r}
load("~/MIR123/mirDriverFold/miRDriver_Step1/GisticResults/gistic_bedfile.rda")
load(system.file("extdata", "hg38_genes_coord_bedfile.rda", package="miRDriver"))
makingCisAndTransGenes(Genes_coord_bedfile = Genes_coord, gistic_bedfile=region, mirdirectory="~/MIR123")
```

Output: The output for this function can be found in the mirdirectory in mirDriverFold/miRDriver_Step2/Trans_Cis_Files. There is a directory with four files in it for each peak. 

```{r}
load("~/MIR123/mirDriverFold/miRDriver_Step2/Trans_Cis_Files/Amplification Peak  1/trans_genes.Rda")
trans_genes[1:3,]
```
```{r}
load("~/MIR123/mirDriverFold/miRDriver_Step2/Trans_Cis_Files/Amplification Peak  1/trans_UPgene_Amplification Peak  1.Rda")
trans_genes[1:3,]
```
```{r}
load("~/MIR123/mirDriverFold/miRDriver_Step2/Trans_Cis_Files/Amplification Peak  1/trans_DOWNgene_Amplification Peak  1.Rda")
trans_genes[1:3,]
```
```{r}
load("~/MIR123/mirDriverFold/miRDriver_Step2/Trans_Cis_Files/Amplification Peak  1/cis_genes.Rda")
cis_genes
```

##<a id = "Step-3"></a> Step 3
Step three of the miRDriver pipeline is the gathering of the cis genes and cis miRNA for each trans gene. This step uses one function, *gatherCisGenes()*.

### <a id = "gatherCisGenes"></a> gatherCisGenes
This function collects all the DE cis genes and all miRNAs for each trans gene. This function uses the output from the *makingCisAndTransGene()* function as well as a *mirna_bedfile* and a *gistic_bedfile*. The gistic_bedfile is described in the makingCisAndTransGene function description and can be made using the *getRegionWiseGistic()* function. The *mirna_bedfile* is described in the example datasets section and needs to be provided by the user. The output from this function is in the folders gene_all_DEcis and gene_all_mirna. In these folders there is a folder for each trans gene where all the miRNA's or DE cis genes are listed. As with all the functions the mirdirectory needs to be consistent with the other functions.

Arguments:

ncores - Assuming parallel, and integer value that represents the number of cores that the user wants to use (default = 2), this argument will be ignored if parallel is set to FALSE

mirna_bedfile - A dataframe in the format of a bedfile for all of the miRNA

gistic_bedfile - A dataframe in the format of a bedfile for the GISTIC results

parallel - A logic value (TRUE/FALSE) for whether or not one wants to run the code in parallel

mirdirectory - a character string for the file path of the directory one wants the results , must be the same as in all previous functions


```{r}
load(system.file("extdata", "mirna_bedfile.Rda", package="miRDriver"))
load("~/MIR123/mirDriverFold/miRDriver_Step1/GisticResults/gistic_bedfile.rda")
dummy_mirna_bedfile<-mirna_bedfile[1:100,]
gatherCisGenes(ncores = 2,mirna_bedfile=dummy_mirna_bedfile,gistic_bedfile=region,parallel = FALSE,mirdirectory="~/MIR123")

```

Output: The output for this function can be found in the mirdirectory in mirDriverFold/miRDriver_Step3. The mirna files for each gene are in the folder gene_all_mirna and the DE cis genes for each gene are in the folder gene_all_DEcis. In this case there are no DECis genes for our output due to the dummy datasets, but the dataframes will follow the pattern of the mirna dataframe.

```{r}
load("~/MIR123/mirDriverFold/miRDriver_Step3/gene_all_DEcis/ENSG00000100242.Rda")
TC
```
```{r}
load("~/MIR123/mirDriverFold/miRDriver_Step3/gene_all_mirna/ENSG00000124205.Rda")
GM[1:3,]
```

##<a id = "Step-4"></a> Step 4
Step four of the miRDriver pipeline is to select potential miRNAs regulating DE trans genes. This step uses two functions, *getTransGenePredictorFile()* and *lassoParallelForTransGene()*.

### <a id = "getTransGenePredictorFile"></a> getTransGenePredictorFile
This function creates a TransGenePredictorFile for each trans gene based on the mirna and DE cis files from the output of the *gatherCisGenes* function and the user provided dataframes *methylationData*, *RSeqData*, *CNVData*, *TFData*, *mirnaData*. The five dataframes needed for this function are described in the example datasets section. The TransGenePredictorFiles will contain information about the possible predictors for each patient based on the data provided. The TransGenePredictorFiles will be saved in the file path ~/mirdirectory/mirDriverFold/miRDriver_Step4/TransGenePredFile/. As always mirdirectory in the file path represents the mirdirectory chosen for the analysis and should be the same as the mirdirectory used in all previous functions.

Arguments:

ncores - assuming parallel use, the number of cores the user wants to use (default=2), this argument will be ignored if parallel=FALSE

methylationData - a dataframe for the methylation data

RSeqData - a dataframe for the RNASeq data

CNVData - a dataframe for the copy number varation data

TFData - a dataframe for the transcription factor data

mirnaData - a dataframe for the mirna data

parallel - a logic value for whether or not the user wants to run the code in parallel

mirdirectory - a character string for the file path of the directory one wants the results in, should be the same as in previous functions

```{r}
load(system.file("extdata", "dummy_methylationData10.Rda", package="miRDriver"))
load(system.file("extdata", "dummy_RSeqData10_200.Rda", package="miRDriver"))
load(system.file("extdata", "dummy_CNVData10.Rda", package="miRDriver"))
load(system.file("extdata", "dummy_TFData.Rda", package="miRDriver"))
load(system.file("extdata", "dummy_mirnaData10.Rda", package="miRDriver"))

getTransGenePredictorFile(ncores = 2,methylationData=dummy_methylationData10,RSeqData=dummy_RSeqData10_200,CNVData=dummy_CNVData10,TFData=dummy_TFData,mirnaData=dummy_mirnaData10,parallel = FALSE,mirdirectory="~/MIR123")

```

Output: The output for this function can be found in the mirdirectory in mirDriverFold/miRDriver_Step4/TransGenePredFile.

```{r}
load("~/MIR123/mirDriverFold/miRDriver_Step4/TransGenePredFile/ENSG00000108231.Rda")
GenePred[1:3,]
```



### <a id = "lassoParallelForTransGene"></a> lassoParallelForTransGene

This function runs a LASSO regression to find what contributes to the expression of the trans gene. This type of regression eliminate non-contributors by settting their coefficient to zero. If the coeffiecient of a miRNA is non-zero after the regression finishes then it is a contributor. However, to be accurate the regression is set to run multiple times (the user specifies as an argument), with a certain percentage of times needed with the coefficient to be non-zero. (the user specifies this as an argument as well). The user also needs to provide the number of folds they want their regression to use. The regression is run on the TransGenePredictorFiles created by the getTransPredictorFile function. The output from this function is found in the file path ~/mirdirectory/mirDriverFold/miRDriver_RESULTS/. The mirdirectory in the file path represents the directory chosen by the user and specified in the argument, mirdirectory. The argument *mirdirectory* needs to be set to the same directory as the mirdirectory in the previous steps. 

Arguments:

ncores - assuming the code in run in parallel, an integer for the number of cores one wants to use (default = 2), if parallel is set to FALSE then it will be ignored

Nfolds - an integer for the number of times the LASSO regression is run

numCounter - an integer for the number of folds one wants for their LASSO regression

nonZeroPercent - an integer for the percentage of appearance as a non-zero LASSO   coefficient for a miRNA to be selected

parallel - a logical value (TRUE/FALSE) for whether or not the user wants to run the code in parallel

mirdirectory - a character string for the file path of the directory one wants the results in

```{r}
lassoParallelForTransGene(ncores = 2,numCounter = 10,Nfolds = 3,nonZeroPercent = 20,parallel = FALSE, mirdirectory = "~/MIR123")
```

Output: The output for this function can be found in the mirdirectory in mirDriverFold/miRDriver_RESULTS. There is a field for each gene and its predictors in the LassoMinCoeff folder and a dataframe with the miRNA that are most significant.

```{r}
load("~/MIR123/mirDriverFold/miRDriver_RESULTS/LassoMinCoeff/ENSG00000088387.Rda")
min_Coeff[1:3,]
```


# <a id = "Summary"></a> Summary

This package is designed to Infer Copy Number Derived miRNA-Gene Networks in Cancer. The four step process breaks up the different parts of the process into chunks for easy use and produces accurate results.

# <a id = "Acknowledgements"></a> Acknowledgements

This work was supported by the National Institute of General Medical Sciences of the National Institutes of Health under Award Number R35GM133657.
